### PR TITLE
Update schedule

### PR DESCRIPTION
**Pull request by gediminasel**
screenshot_2016-08-28-12-40-43 1

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -19,4 +18,5 @@
     <string name="users">Dalyviai</string>
     <string name="timer">Laikrodukas</string>
     <string name="schedule">Programa</string>
+    <string name="old_data_in_API">API\'uje nėra naujų duomenų.</string>
```
Comment (Dekedro) [inline app/src/main/res/values/strings.xml]: It should be more user friendly

File: app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java [hunk 2]
```diff
@@ -84,9 +84,25 @@ private void setScheduleItems(List<ScheduleEvent> scheduleEvents) {
...
-        ScheduleAdapter adapter = new ScheduleAdapter(getContext(), scheduleItems);
+        ScheduleAdapter adapter = new ScheduleAdapter(scheduleItems);
         scheduleListView.setAdapter(adapter);
+        scheduleListView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
+        scheduleListView.setHasFixedSize(false);
+        if (i < scheduleItems.size()) {
+            scheduleListView.scrollToPosition(i);
+            //TODO if(i+1==scheduleItems.size())    bottomnavigation.dontshow
+        } else {
+            Error.oldData(getView());
```
Comment (Dekedro) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java]: WTF? If there is no upcoming events, then you shouldn't show error or do anything like this. Why did you add this functionality?

File: app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java [hunk 3]
```diff
@@ -84,9 +84,25 @@ private void setScheduleItems(List<ScheduleEvent> scheduleEvents) {
...
 
         Collections.sort(scheduleItems, new ScheduleItemComparator());
+        int i = 0;
+        Date now = new Date();
+        while (i < scheduleItems.size()) {
+            if (scheduleItems.get(i).getDate().before(now))
+                i = i + 1;
+            else
+                break;
+        }
```
Comment (Dekedro) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java]: Something like this could save 5 lines of code.

int i = 0;
Date now = new Date();
for (i = 0; i < scheduleItems.size () && !scheduleItems.get(i).getDate().before(now); ++i)

File: app/src/main/res/layout/schedule_day.xml [hunk 4]
```diff
@@ -1,13 +0,0 @@
...
-    android:orientation="horizontal">
-
-    <TextView
-        android:id="@+id/schedule_time"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_gravity="center"
-        android:text="Hello World!"
-        android:textSize="18dp" />
-</LinearLayout>
```
Comment (Dekedro) [inline app/src/main/res/layout/schedule_day.xml:13]: Why? I don't want to go deeper into code, so could you explain your changes briefly? Did you remove schedule_day?

Comment (gediminasel) [inline app/src/main/res/layout/schedule_day.xml]: I created styles schedule_day and schedule_item_name

**Comment by gediminasel** [comment]
Laukiu komentarų

File: app/src/main/java/com/nmakademija/nmaakademija/adapter/ScheduleAdapter.java [hunk 5]
```diff
@@ -1,69 +1,74 @@
...
+            endTime = (TextView) view.findViewById(R.id.schedule_end_time);
+            name = (TextView) view.findViewById(R.id.schedule_name);
+            author = (TextView) view.findViewById(R.id.schedule_author);
         }
-        return convertView;
     }
 
-    private View getViewScheduleEvent (ScheduleEvent scheduleEvent, View convertView, ViewGroup parent){
-        convertView = LayoutInflater.from(getContext()).inflate(R.layout.schedule_event, parent, false);
+    private Context _context;
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/adapter/ScheduleAdapter.java]: Don't start variable names with underscore. Camel case is good enough in these cases. More about android guidelines https://github.com/ribot/android-guidelines/blob/master/project_and_code_guidelines.md 
P.S I wouldn't recommend to add `m` and `s` prefixes to variable names.

File: app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java [hunk 6]
```diff
@@ -85,8 +85,28 @@ private void setScheduleItems(List<ScheduleEvent> scheduleEvents) {
...
 
+        int i;
+        Date now = new Date();
+        for (i = 0; i < scheduleItems.size() && scheduleItems.get(i).getDate().before(now); ++i) ;
+
         ScheduleAdapter adapter = new ScheduleAdapter(getContext(), scheduleItems);
-        scheduleListView.setAdapter(adapter);
+        scheduleRecyclerView.setAdapter(adapter);
+        scheduleRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
+        scheduleRecyclerView.setHasFixedSize(false);
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java]: Default is false

File: app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java [hunk 7]
```diff
@@ -85,8 +85,28 @@ private void setScheduleItems(List<ScheduleEvent> scheduleEvents) {
...
         Collections.sort(scheduleItems, new ScheduleItemComparator());
 
+        int i;
+        Date now = new Date();
+        for (i = 0; i < scheduleItems.size() && scheduleItems.get(i).getDate().before(now); ++i) ;
+
         ScheduleAdapter adapter = new ScheduleAdapter(getContext(), scheduleItems);
-        scheduleListView.setAdapter(adapter);
+        scheduleRecyclerView.setAdapter(adapter);
+        scheduleRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java]: You can set `layoutManager` in xml layout file with `app:layoutManager` property

File: app/src/main/res/layout/activity_schedule.xml [hunk 8]
```diff
@@ -10,8 +11,10 @@
     android:paddingTop="@dimen/activity_vertical_margin"
     tools:context=".fragment.ScheduleFragment">
 
-    <ListView
+    <android.support.v7.widget.RecyclerView
```
Comment (vycius) [inline app/src/main/res/layout/activity_schedule.xml]: `FrameLayout` is useless here. Leave only `RecyclerView` here.

File: app/src/main/res/values/styles.xml [hunk 9]
```diff
@@ -8,4 +8,14 @@
         <item name="colorAccent">@color/colorAccent</item>
     </style>
 
+    <style name="Schedule_day">
```
Comment (vycius) [inline app/src/main/res/values/styles.xml]: There is some built in text apperances from AppCompat. See https://gist.github.com/Aracem/922d460b73168676ca21 parents. I hope they will be useful.

File: app/build.gradle [hunk 10]
```diff
@@ -7,8 +7,8 @@ android {
         applicationId "com.nmakademija.nmaakademija"
         minSdkVersion 16
         targetSdkVersion 24
-        versionCode 1
-        versionName "1.0"
+        versionCode 101
```
Comment (kuprelis) [inline app/build.gradle]: cia turetu but 2

Comment (gediminasel) [inline app/build.gradle]: Kodėl?

File: app/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -1,6 +1,5 @@
 <resources>
     <string name="app_name">NMAkademija</string>
-    <string name="loading">Loading...</string>
```
Comment (vycius) [inline app/src/main/res/values/strings.xml:3]: This is still used in `activity_time_until_session.xml` 22 line.

Comment (vycius) [inline app/src/main/res/values/strings.xml:3]: You don't get error, because it's in `tools:text` property

File: app/src/main/java/com/nmakademija/nmaakademija/adapter/ScheduleAdapter.java [hunk 12]
```diff
@@ -1,69 +1,74 @@
...
+            holder.startTime.setVisibility(View.VISIBLE);
+            holder.author.setVisibility(View.VISIBLE);
+            holder.name.setTextAppearance(context, R.style.Schedule_item_name);
+        } else {
+            ScheduleDayBanner scheduleDayBanner = (ScheduleDayBanner) scheduleItem;
+            holder.name.setTextAppearance(context, R.style.Schedule_day);
+            holder.name.setText(scheduleDayBanner.getTime());
+            holder.endTime.setVisibility(View.GONE);
+            holder.startTime.setVisibility(View.GONE);
+            holder.author.setVisibility(View.GONE);
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/adapter/ScheduleAdapter.java]: Use two different `ViewHolders`. More about it here http://stackoverflow.com/questions/26245139/how-to-create-recyclerview-with-multiple-view-type

File: app/src/main/java/com/nmakademija/nmaakademija/adapter/ScheduleAdapter.java [hunk 13]
```diff
@@ -1,69 +1,74 @@
...
-    public ScheduleAdapter(Context context, ArrayList<ScheduleItem> events){
-        super(context, R.layout.schedule_event, events);
-    }
+import java.util.List;
 
-    @Override
-    public boolean isEnabled (int position){
-        return false;
-    }
+public class ScheduleAdapter extends RecyclerView.Adapter<ScheduleAdapter.MyViewHolder> {
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/adapter/ScheduleAdapter.java]: Nice code example about Recycler view with sections https://gist.github.com/gabrielemariotti/4c189fb1124df4556058

**Comment by gediminasel** [comment]
screenshot_2016-08-27-21-33-20 1

File: app/src/main/java/com/nmakademija/nmaakademija/MainActivity.java [hunk 14]
```diff
@@ -32,10 +32,19 @@ protected void onCreate(Bundle savedInstanceState) {
...
-        setCurrentFragment(0);
+
+        selectedIndex = 0;
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        setCurrentFragment(selectedIndex);
+        bottomNavigation.setSelectedIndex(selectedIndex, false);
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/MainActivity.java]: Why do we need this in `onResume` ?

File: app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java [hunk 15]
```diff
@@ -85,8 +85,27 @@ private void setScheduleItems(List<ScheduleEvent> scheduleEvents) {
...
+
         ScheduleAdapter adapter = new ScheduleAdapter(getContext(), scheduleItems);
-        scheduleListView.setAdapter(adapter);
+        scheduleRecyclerView.setAdapter(adapter);
+//        scheduleRecyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
+//
+        scheduleRecyclerView.scrollToPosition(Math.min(i, scheduleItems.size() - 1));
+
+//        scheduleRecyclerView.smoothScrollToPosition(Math.min(i, scheduleItems.size() - 1));
+//        scheduleRecyclerView.getLayoutManager().smoothScrollToPosition(scheduleRecyclerView, null, scheduleItems.size() - 1);
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java]: Remove comments

File: app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java [hunk 16]
```diff
@@ -70,13 +71,14 @@ public void onFailure(Call<List<ScheduleEvent>> call, Throwable t) {
     }
 
     private void setScheduleItems(List<ScheduleEvent> scheduleEvents) {
-        ArrayList<ScheduleItem> scheduleItems = new ArrayList<>();
+        /*ArrayList<ScheduleItem> scheduleItems = new ArrayList<>();
         scheduleItems.addAll(scheduleEvents);
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java]: Remove comments

File: app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java [hunk 17]
```diff
@@ -70,23 +71,34 @@ public void onFailure(Call<List<ScheduleEvent>> call, Throwable t) {
...
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.US);
         ArrayList<Date> scheduleDates = new ArrayList<>();
-        for (ScheduleEvent scheduleEvent : scheduleEvents) {
-            Date eventDate = scheduleEvent.getDate();
+        List<ScheduleSectionsAdapter.Section> sections = new ArrayList<>();
+        for (int i = 0; i < scheduleEvents.size(); i++) {
+            Date eventDate = scheduleEvents.get(i).getDate();
             Date date = new Date(eventDate.getYear(), eventDate.getMonth(), eventDate.getDate());
+
             if (!scheduleDates.contains(date)) {
```
Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java:84]: `contains()` complexity on `ArrayList` is `O(N)`. Switch to `HashSet` to have `O(1)` complexity `insert` and `contains`

Comment (vycius) [inline app/src/main/java/com/nmakademija/nmaakademija/fragment/ScheduleFragment.java:84]: We can change this in separate PR.

**Comment by vycius** [comment]
👍 LGTM nice work

File: app/src/main/java/com/nmakademija/nmaakademija/adapter/ArticlesAdapter.java [hunk 18]
```diff
@@ -64,4 +51,16 @@ public void onClick(View view) {
     public int getItemCount() {
         return articlesList.size();
     }
+
+    public class MyViewHolder extends RecyclerView.ViewHolder {
```
Comment (ignas-gustainis) [inline app/src/main/java/com/nmakademija/nmaakademija/adapter/ArticlesAdapter.java:55]: It would be good to use more informative class names, so anyone that looks at code can instantly know the purpose of the class

File: app/src/main/java/com/nmakademija/nmaakademija/entity/UserRoles.java [hunk 19]
```diff
@@ -6,9 +6,9 @@
  * Created by dekedro on 16.8.24.
  */
 public enum UserRoles {
-    @SerializedName("teacher") TEACHER,
```
Comment (ignas-gustainis) [inline app/src/main/java/com/nmakademija/nmaakademija/entity/UserRoles.java:9]: Why the spaces removal in this file?

**Comment by ignas-gustainis** [comment]
Besides the comments I wrote, LGTM :+1:
Good job

File: app/src/main/java/com/nmakademija/nmaakademija/entity/UserRoles.java (hunk 19 above)
Comment (vycius) [reply app/src/main/java/com/nmakademija/nmaakademija/entity/UserRoles.java:9]: Most of the time we write annotations over variable such as:

``` java
    @SerializedName("teacher")
    TEACHER,

    @SerializedName("member")
    MEMBER,

    @SerializedName("guest")
    GUEST,

    @SerializedName("supervisor")
    SUPERVISOR,

    @SerializedName("administrator")
    ADMINISTRATOR
```

Although this syntax is also correct. I am not sure about space removal. I think `Android studio` wanted to move `MEMBER`, `GUEST` to separate line.

**Comment by gediminasel** [comment]
I am merging it in the evening
